[rewrite] Reimplement Gobbie Bag quests using Interaction Framework

Kietsu3756 opened the pull request:
Converted "The Gobbiebag Part *" quest line to Interaction Framework.

1. Added "Bowl of Goblin stew 880" to trade list even though it may be unobtainable w/o GM commands currently
2. To reduce redundant code, used a derived GobbieQuest class/meta table from Quest
3. Used fame values from https://www.bg-wiki.com/ffxi/The_Gobbiebag_Part_I ... https://www.bg-wiki.com/ffxi/The_Gobbiebag_Part_X
4. Tested turn in each item individually and the stew with GM commands
5. Added a GM command to ease testing and to set bag size

Kietsu3756 committed:
Update xi.items table for Gobbiebag quest line, update Ghatsad with SQUARE_OF_VELVET_CLOTH rename

Kietsu3756 committed:
Remove Gobbiebag quest logic from Bluffnix NPC script

Kietsu3756 committed:
Interaction - The Gobbiebag Part I

Added GobbieQuest derived class from Quest to be used by follow up quest line

Kietsu3756 committed:
Interaction - The Gobbiebag Part II - X

Each follow up Gobbiebag quest uses the GobbieQuest derived class from Quest

Kietsu3756 committed:
Add Bluffnix default action

Kietsu3756 committed:
Add !setbag GM command

zach2good commented:
Looks pretty good at a glance, I like the quest baseclass! I'll leave the main review to someone else
(I updated the title to be prettier for the changelogs)

claywar commented on scripts/quests/jeuno/The_Gobbiebag_Part_I.lua:
```diff
@@ -0,0 +1,131 @@
+setmetatable(GobbieQuest, {__index = Quest} )
+GobbieQuest.__index = GobbieQuest
+
+function GobbieQuest:new(params)
```
I feel like the better place to store the base class would be in a helpers.lua file.  It feels a bit awkward for later quests to be calling in the initial quest script.

Kietsu3756 replied:
Yeah... to be honest I have been playing around with this for a while before I finally got it to work properly. Agree with you on the awkwardness factor, I was hoping you guys would tell me a good place to put it :-) Sounds good, l will move and make sure everything works as expected.

claywar commented on scripts/quests/jeuno/The_Gobbiebag_Part_I.lua:
```diff
@@ -0,0 +1,131 @@
+    local getPendingDialogueId = function (player)
+        return (player:getContainerSize(xi.inv.INVENTORY) + 1)
+    end
+    -- If quest is completed, the corect dialogue ID is the expected post quest inventory size offset by 1
```
Just a styling thing, could you please add a newline between functions?

Kietsu3756 replied:
Sure, np, also I think I forgot to add , to the last entry in all of the the param tables.

claywar reviewed (changes requested):
In all, this is beautiful work 👍 .  My main request is for moving common procedures out of the first quest, and into a generic helpers file (which appears to be currently supported for quests as well in luautils).  Other than that, a few styling things, and its good to go.  Very nice again!

Kietsu3756 commented:
@zach2good, @claywar: Thank you both for the comments and review.

I retested using using the stew with the gm commands for all quests + expected default text once complete. I only tested Gobbiebag I and Gobbiebag X with individual item requests since I did not modify any of the tradeItems tables (except for rename).

It was fun, I learned some Lua and hope to contribute more in the future.

claywar commented:
I apologize for the confusion.  When I said helpers, I really meant a file called helpers.lua, and not attaching to utils.  We have support in each sub-directory for quests and missions for a specific file to handle that.  For an example, `missions/amk` has one implemented there.  Using a similar format, I think that would be the best place to put the gobbie bag quest common code.

Kietsu3756 committed:
Refactor per review comments

- helpers.lua
   - Add common requires list
   - Make GobbieQuest a part of xi.jeuno.helpers and rename to GobbiebagQuest
   - Add linefeed after local functions
   - Add params.tradeStew, params.message
   - Rename params.trade to params.tradeItems
   - Fix type (corect --> correct)
- The_Gobbiebag_Part_*.lua
   - Adjusted require list
   - Add missing , to last table entry in the parameter table
   - Add tradeItems, tradeStew, and message to parameters tables and refactor spacing

Kietsu3756 committed:
Verbatim move of GobbieQuest base class to new scripts/quests/jeuno/helpers.lua

Kietsu3756 commented:
> I apologize for the confusion. When I said helpers, I really meant a file called helpers.lua, and not attaching to utils. We have support in each sub-directory for quests and missions for a specific file to handle that. For an example, `missions/amk` has one implemented there. Using a similar format, I think that would be the best place to put the gobbie bag quest common code.

No problem, done! Thanks for answering my questions on Discord as well :-)

claywar commented on scripts/quests/jeuno/The_Gobbiebag_Part_I.lua:
```diff
@@ -0,0 +1,36 @@
+
+local quest = xi.jeuno.helpers.GobbiebagQuest:new(params)
+
+return quest
```
Could you please add a newline at the end of file?

Kietsu3756 replied:
I thought my rewrite of the history was flawless... sneaky newline... :-)

claywar reviewed (commented):
One last minor change.  Thanks for being patient and asking questions!

Kietsu3756 committed:
Add missing newline caused by refactor/squash at end of file
